src/stdlib_child_list.f90

```diff
@@ -0,0 +1,369 @@
+!> Implementation of a Child list type to hold various types of data.
+!>
+!> The child list module provides a heterogeneous generic linked list 
+!> that acts as a basic building block for the linked list module
+
+
+module stdlib_child_list
+    implicit none
+
+    ! making Node and child_list struct globally available
+    public:: Node, child_list
+
+    !> Defining Node
+    !>
+    !> The purpose of this node is to hold an item
+    !> and links to previous and next Node.
+    type Node
+        type(Node), pointer :: next => null()
+        type(Node), pointer :: prev => null()
+        class(*), allocatable :: item
+        contains
+        procedure :: clear => node_destroyed
+        procedure, private :: clear_all => all_nodes_destroyed
+    end type Node
+
+    !> Defining Child List
+    !>
+    !> This linked list is single-dimensional chain of Nodes.
+    !> It is a doubly-linked heterogeneous generic list .
+    type child_list
+        integer, private :: num_nodes = 0
+        type(Node), pointer :: head => null()
+        type(Node), pointer :: tail => null()
+        contains
+        procedure:: push => push_at_tail
+        procedure:: insert => insert_at_index
+        procedure:: pop => pop_node_at_tail
+        procedure:: remove => remove_node_at_index
+        procedure:: get => get_node_at_index
+        procedure:: size => get_length
+        procedure:: set_size => set_length
+        procedure:: replace => replace_at_index
+        procedure:: reverse => reverse_child_list
+        procedure:: clear => destroy_whole_child_list
+    end type child_list
+
+    contains
+
+    !> Creates a Node that contains 'new_item' as its child
+    !>
+    !> Returns the new parent node   
+    pure function initialize_node( new_item ) result( new_node )
+        type(node) :: new_node
+        class(*), intent(in), optional :: new_item
+        
+        ! allocating new_item to the new node's item
+        allocate(new_node%item, source=new_item)
```

Comment:
`source` uses the optional argument `new_item` here, there may be memory access errors.